src/main/kotlin/com/springernature/newversion/BuildpackVersionChecker.kt

```diff
@@ -4,28 +4,52 @@ import org.slf4j.Logger
 import org.slf4j.LoggerFactory
 import java.io.File
 
+
+sealed class ChecksResult {
+    abstract val errors: Map<BuildpackUpdate, Exception>
+}
+
+data class SuccessfulChecks(val updates: List<BuildpackUpdate>) : ChecksResult() {
+    override val errors: Map<BuildpackUpdate, Exception> = emptyMap()
+}
+
+data class FailedChecks(val updates: List<BuildpackUpdate>, override val errors: Map<BuildpackUpdate, Exception>) :
+    ChecksResult()
+
 class BuildpackVersionChecker(
     private val manifestPath: File,
     private val buildpackUpdateChecker: BuildpackUpdateChecker,
     private val publisher: Publisher
 ) {
 
-    fun performChecks() {
-        ManifestParser.load(manifestPath)
+    fun performChecks(): ChecksResult {
+        LOG.info("Performing checks")
+        val errors = LinkedHashMap<BuildpackUpdate, Exception>()
+        val updates: List<BuildpackUpdate> = ManifestParser.load(manifestPath)
             .flatMap { ManifestBuildpack.from(it) }
             .filter { it.buildpack.version != Unparseable }
             .groupBy { it.buildpack }
             .map { (buildpack, manifestBuildpacks) ->
-                BuildpackUpdate(manifestBuildpacks.map { it.manifest }, buildpack, buildpackUpdateChecker.findLatestVersion(buildpack))
+                BuildpackUpdate(
+                    manifestBuildpacks.map { it.manifest },
+                    buildpack,
+                    buildpackUpdateChecker.findLatestVersion(buildpack)
+                )
             }
             .filter(BuildpackUpdate::hasUpdate)
+        updates
```

Comment:
Why do we break the chain at this point?

Reply:
Because so far I only knew `forEach` (which returns `Unit`). Now I found [`onEach`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.sequences/on-each.html). :smile_cat: